### DISCUSSION TITLE
With-expression support for copyable classes

### DISCUSSION THREAD

**Discussion by bernd5** [dotnet/csharplang#4514] [Q&A]
I have seen that there are proposals to extend the with-expression introduced with records to structs - but are there plans to allow with expressions for regular classes, too?

Of course such a class would need a parameterless Copy method and the correct return type.

**Comment by ufcpp** [comment]
see https://github.com/dotnet/csharplang/blob/a3d67d141501101af9dbb9e8192ca8b4c717e316/meetings/2020/LDM-2020-04-20.md

> We'll consider broader Factory scenarios later.

**Reply by bernd5** [reply]
The extension to factory methods is a nice idea - but in my eyes it should be extended to copyable classes which have a parameterless method named "Copy" and a matching return type or at least for classes which implement `ICloneable`, too.

**Reply by ufcpp** [reply]
```cs
var x = new X { Value = 1 };
var y = x with { Value = 2 }; // error with C# 9.0 but ok with this proposal

Console.WriteLine(x.Value); // 2?

class X : ICloneable
{
    public int Value { get; init; }

    // ICloneable interface has no guarantee of creating new instance
    public object Clone() => this;
}
```

**Reply by bernd5** [reply]
Then you violate the ICloneable contract. I mean you could for example overload operator+ and implement there some substraction - this is crazy but no reason to disallow operator overloading.

But I understand such an implementation is intended as "Clone-optimization" which is perfectly legal as long as X and all of its child classes are immutable.
